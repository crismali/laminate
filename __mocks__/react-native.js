/* eslint-env jest */

class Value {
  constructor (value) {
    this.value = value
  }

  interpolate (args) {
    return { interpolated: this.value, args: args }
  }
}

const AnimatedView = ({ children }) => children
AnimatedView.displayName = 'Animated.View'

const animatedTimingStart = jest.fn()
const Animated = {
  Value,
  View: AnimatedView,
  timing: jest.fn(() => ({ start: animatedTimingStart })),
  animatedTimingStart
}

const Dimensions = { get: jest.fn(() => ({ width: 375, height: 667 })) }

const PanResponder = { create: jest.fn((args) => ({ panResponder: args })) }

const View = ({ children }) => children

const StyleSheet = { create: jest.fn(styles => styles) }

export {
  Animated,
  Dimensions,
  PanResponder,
  StyleSheet,
  View
}
