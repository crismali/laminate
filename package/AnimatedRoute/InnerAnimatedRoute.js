import { Component } from 'react'
import { Animation } from '../Animation'

export class InnerAnimatedRoute extends Component {
  static propTypes = {
    ...Animation.childContextTypes
  }

  state = {
    children: this.props.children
  }

  componentDidMount () {
    if (this.props.children) {
      this.props.animation.play()
    }
  }

  componentWillReceiveProps (nextProps) {
    const { children: currentChildren } = this.props
    const { children: nextChildren } = nextProps

    if (this.isEntering(currentChildren, nextChildren)) {
      this.enter(nextChildren)
    } else if (this.isExiting(currentChildren, nextChildren)) {
      this.exit()
    } else {
      this.setState({ children: nextChildren })
    }
  }

  enter = (children) => {
    this.setState({ children }, this.props.animation.play)
  }

  exit = () => {
    this.props.animation.rewind({ callback: this.clearChildren })
  }

  clearChildren = () => this.setState({ children: null })

  isEntering = (currentChildren, nextChildren) => {
    return !currentChildren && nextChildren
  }

  isExiting = (currentChildren, nextChildren) => {
    return currentChildren && !nextChildren
  }

  render () {
    return this.state.children
  }
}
