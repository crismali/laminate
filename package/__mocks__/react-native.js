/* eslint-env jest */
import React from 'react'

class Value {
  constructor (value) {
    this.value = value
  }

  interpolate (args) {
    return { interpolated: this.value, args: args }
  }

  setValue (value) {
    this.value = value
  }
}

const AnimatedView = ({ children }) => children
AnimatedView.displayName = 'Animated.View'

const animatedTimingStart = jest.fn((func) => func && func())
const Animated = {
  Value,
  View: AnimatedView,
  timing: jest.fn(() => ({ start: animatedTimingStart })),
  animatedTimingStart
}

const Dimensions = { get: jest.fn(() => ({ width: 375, height: 667 })) }

const Easing = { inOut: jest.fn() }

const PanResponder = { create: jest.fn((args) => ({ panHandlers: args })) }

const View = ({ children }) => <div>{children}</div> // eslint-disable-line

const StyleSheet = { create: jest.fn(styles => styles) }

export {
  Animated,
  Dimensions,
  Easing,
  PanResponder,
  StyleSheet,
  View
}
