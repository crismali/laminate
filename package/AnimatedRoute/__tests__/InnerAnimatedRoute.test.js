/* eslint-env jest */
import React from 'react'
import { Animated } from 'react-native'
import { shallow } from 'enzyme'
import { InnerAnimatedRoute } from '../InnerAnimatedRoute'

const Child = () => 'a child'

const animationMock = () => {
  return {
    play: jest.fn(),
    rewind: jest.fn(),
    value: new Animated.Value(17)
  }
}

describe('state', () => {
  it('has children match the children in props', () => {
    const children = <Child />
    const subject = shallow(<InnerAnimatedRoute animation={animationMock()}>
      {children}
    </InnerAnimatedRoute>)
    expect(subject).toHaveState('children', children)
  })
})

describe('#componentDidMount', () => {
  it('plays the animation when given children', () => {
    const animation = animationMock()
    const subject = shallow(<InnerAnimatedRoute animation={animation}>
      <Child />
    </InnerAnimatedRoute>)
    subject.instance().componentDidMount()
    expect(animation.play).toHaveBeenCalled()
  })

  it('does not play the animation when there are no children', () => {
    const animation = animationMock()
    const subject = shallow(<InnerAnimatedRoute animation={animation}>
      {null}
    </InnerAnimatedRoute>)
    subject.instance().componentDidMount()
    expect(animation.play).not.toHaveBeenCalled()
  })
})

describe('#componentWillReceiveProps', () => {
  describe('when entering', () => {
    it('enters', () => {
      const subject = shallow(<InnerAnimatedRoute animation={animationMock()}>
        {null}
      </InnerAnimatedRoute>)
      subject.instance().enter = jest.fn()
      subject.instance().exit = jest.fn()
      subject.instance().setState = jest.fn()
      subject.setProps({ animation: animationMock(), children: <Child /> })

      expect(subject.instance().enter).toHaveBeenCalledWith(<Child />)
      expect(subject.instance().exit).not.toHaveBeenCalled()
      expect(subject.instance().setState).not.toHaveBeenCalled()
    })
  })

  describe('when exiting', () => {
    it('exits', () => {
      const subject = shallow(<InnerAnimatedRoute animation={animationMock()}>
        <Child />
      </InnerAnimatedRoute>)
      subject.instance().enter = jest.fn()
      subject.instance().exit = jest.fn()
      subject.instance().setState = jest.fn()
      subject.setProps({ animation: animationMock(), children: null })

      expect(subject.instance().exit).toHaveBeenCalled()
      expect(subject.instance().enter).not.toHaveBeenCalled()
      expect(subject.instance().setState).not.toHaveBeenCalled()
    })
  })

  describe('when not entering and not exiting', () => {
    it('updates the children', () => {
      const subject = shallow(<InnerAnimatedRoute animation={animationMock()}>
        <Child foo='bar' />
      </InnerAnimatedRoute>)
      subject.instance().enter = jest.fn()
      subject.instance().exit = jest.fn()
      subject.instance().setState = jest.fn()
      subject.setProps({ animation: animationMock(), children: <Child /> })

      expect(subject.instance().setState).toHaveBeenCalledWith({ children: <Child /> })
      expect(subject.instance().exit).not.toHaveBeenCalled()
      expect(subject.instance().enter).not.toHaveBeenCalled()
    })
  })
})

describe('#enter', () => {
  it('updates the children and plays the animation afterwards', () => {
    const animation = animationMock()
    const subject = shallow(<InnerAnimatedRoute animation={animation}>
      {null}
    </InnerAnimatedRoute>)
    subject.instance().setState = jest.fn()
    subject.instance().enter(<Child />)
    expect(subject.instance().setState).toHaveBeenCalledWith({ children: <Child /> }, animation.play)
  })
})

describe('#exit', () => {
  it('rewinds the animation and updates the children afterwards', () => {
    const animation = animationMock()
    const subject = shallow(<InnerAnimatedRoute animation={animation}>
      <Child />
    </InnerAnimatedRoute>)
    subject.instance().exit()
    expect(animation.rewind).toHaveBeenCalledWith({ callback: subject.instance().clearChildren })
  })
})

describe('#clearChildren', () => {
  it('sets children to null', () => {
    const subject = shallow(<InnerAnimatedRoute animation={animationMock()}>
      <Child />
    </InnerAnimatedRoute>)
    subject.instance().clearChildren()
    expect(subject).toHaveState('children', null)
  })
})

describe('#render', () => {
  it('renders the children in state', () => {
    const subject = shallow(<InnerAnimatedRoute animation={animationMock()}>
      <Child foo='bar' />
    </InnerAnimatedRoute>)
    subject.setState({ children: <Child new='yes' /> })
    expect(subject).toMatchSnapshot()
  })
})
