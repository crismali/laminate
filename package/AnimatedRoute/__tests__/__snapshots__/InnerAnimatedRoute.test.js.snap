// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render renders the children in state 1`] = `
<Child
  new="yes"
/>
`;
